pkg/operator2/oauth.go

```diff
@@ -136,3 +148,173 @@ func (c *osinOperator) handleOAuthConfig(configOverrides []byte) (*corev1.Config
+		}
+		serializer.Encode(&p, bytesWriter)
+
+	case configv1.IdentityProviderTypeGitHub:
+		githubConfig := providerConfig.GitHub
+		p := osinv1.GitHubIdentityProvider{
+			ClientID: githubConfig.ClientID,
+			ClientSecret: configv1.StringSource{
+				// TODO: this could be Env, Value, File, either can be encrypted
+				StringSourceSpec: configv1.StringSourceSpec{
+					File: getLocalSecretFilename(githubConfig.ClientSecret),
+				},
+			},
+			Organizations: githubConfig.Organizations,
+			Hostname:      githubConfig.Hostname,
+			CA:            getLocalConfigMapFilename(githubConfig.CA),
+		}
+		serializer.Encode(&p, bytesWriter)
+
+	case configv1.IdentityProviderTypeGitLab:
+		gitlabConfig := providerConfig.GitLab
+		p := osinv1.GitLabIdentityProvider{
+			CA:       getLocalConfigMapFilename(gitlabConfig.CA),
+			URL:      gitlabConfig.URL,
+			ClientID: gitlabConfig.ClientID,
+			ClientSecret: configv1.StringSource{
+				// TODO: this could be Env, Value, File, either can be encrypted
+				StringSourceSpec: configv1.StringSourceSpec{
+					File: getLocalSecretFilename(gitlabConfig.ClientSecret),
+				},
+			},
+			Legacy: &gitlabConfig.LegacyOAuth2, // TODO: better mapping of 2 to 3 values?
+		}
+		serializer.Encode(&p, bytesWriter)
+
+	case configv1.IdentityProviderTypeGoogle:
+		googleConfig := providerConfig.Google
+		p := osinv1.GoogleIdentityProvider{
+			ClientID: googleConfig.ClientID,
+			ClientSecret: configv1.StringSource{
+				// TODO: this could be Env, Value, File, either can be encrypted
+				StringSourceSpec: configv1.StringSourceSpec{
+					File: getLocalSecretFilename(googleConfig.ClientSecret),
+				},
+			},
+			HostedDomain: googleConfig.HostedDomain,
+		}
+		serializer.Encode(&p, bytesWriter)
+
+	case configv1.IdentityProviderTypeOpenID:
+		openIDConfig := providerConfig.OpenID
+		p := osinv1.OpenIDIdentityProvider{
+			CA:       getLocalConfigMapFilename(openIDConfig.CA),
+			ClientID: openIDConfig.ClientID,
+			ClientSecret: configv1.StringSource{
+				// TODO: this could be Env, Value, File, either can be encrypted
+				StringSourceSpec: configv1.StringSourceSpec{
+					File: getLocalSecretFilename(openIDConfig.ClientSecret),
+				},
+			},
+			ExtraScopes:              openIDConfig.ExtraScopes,
+			ExtraAuthorizeParameters: openIDConfig.ExtraAuthorizeParameters,
+			URLs: osinv1.OpenIDURLs{ // TODO: do we need this to be different from configv1?
+				Authorize: openIDConfig.URLs.Authorize,
+				Token:     openIDConfig.URLs.Token,
+				UserInfo:  openIDConfig.URLs.UserInfo,
+			},
+			Claims: osinv1.OpenIDClaims{ // TODO: do we need this to be different from configv1?
+				ID:                []string{"sub"}, // TODO: this is missing in configv1 version?
+				PreferredUsername: openIDConfig.Claims.PreferredUsername,
+				Name:              openIDConfig.Claims.Name,
+				Email:             openIDConfig.Claims.Email,
+			},
+		}
+		serializer.Encode(&p, bytesWriter)
+	} // switch
+
+	bytesWriter.Flush()
```

Comment:
`defer` this at the top.

Reply:
You need to do the flushing before you return, otherwise the bytes buffer to which the `Writer` is connected won't get filled on return.
`defer` would perform the flush after the return is done. I suspect (but did not check the implementation) that `providerConfigBytes.Bytes()` returns a copy of its byte slice rather than the pointer to its internal slice, which would mean that if we deferred the `Flush()`, we'd always return an empty slice (unless we reached the default `Writer` buffer size - 4096.

https://play.golang.org/p/TfhNxdUrYjS